### PR TITLE
[RDY] startup: v:progpath fallback

### LINKED ISSUE #6734: Building fails if procfs is unmounted/incomplete
- Operating system/version: Linux

### Actual behaviour
It fails at this:
```
[100%] Built target printargs-test
make -f runtime/CMakeFiles/vimball-tags.dir/build.make runtime/CMakeFiles/vimball-tags.dir/depend
make[2]: Entering directory `/usr/src/RPM/BUILD/neovim-0.2.0/BUILD'
cd /usr/src/RPM/BUILD/neovim-0.2.0/BUILD && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /usr/src/RPM/BUILD/neovim-0.2.0 /usr/src/RPM/BUILD/neovim-0.2.0/runtime /usr/src/RPM/BUILD/neovim-0.2.0/BUILD /usr/src/RPM/BUILD/neovim-0.2.0/BUILD/runtime /usr/src/RPM/BUILD/neovim-0.2.0/BUILD/runtime/CMakeFiles/vimball-tags.dir/DependInfo.cmake --color=
Dependee "/usr/src/RPM/BUILD/neovim-0.2.0/BUILD/runtime/CMakeFiles/vimball-tags.dir/DependInfo.cmake" is newer than depender "/usr/src/RPM/BUILD/neovim-0.2.0/BUILD/runtime/CMakeFiles/vimball-tags.dir/depend.internal".
Dependee "/usr/src/RPM/BUILD/neovim-0.2.0/BUILD/runtime/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/usr/src/RPM/BUILD/neovim-0.2.0/BUILD/runtime/CMakeFiles/vimball-tags.dir/depend.internal".
Scanning dependencies of target vimball-tags
make[2]: Leaving directory `/usr/src/RPM/BUILD/neovim-0.2.0/BUILD'
make -f runtime/CMakeFiles/vimball-tags.dir/build.make runtime/CMakeFiles/vimball-tags.dir/build
make[2]: Entering directory `/usr/src/RPM/BUILD/neovim-0.2.0/BUILD'
cd /usr/src/RPM/BUILD/neovim-0.2.0/BUILD/runtime/pack/dist/opt/vimball && /usr/bin/cmake -E copy_directory /usr/src/RPM/BUILD/neovim-0.2.0/runtime/pack/dist/opt/vimball /usr/src/RPM/BUILD/neovim-0.2.0/BUILD/runtime/pack/dist/opt/vimball
cd /usr/src/RPM/BUILD/neovim-0.2.0/BUILD/runtime/pack/dist/opt/vimball && /usr/src/RPM/BUILD/neovim-0.2.0/BUILD/bin/nvim -u NONE -i NONE -e --headless -c helptags\ doc -c quit
E685: Internal error: init_path()make[2]: *** [runtime/CMakeFiles/vimball-tags] Error 1
make[2]: Leaving directory `/usr/src/RPM/BUILD/neovim-0.2.0/BUILD'
make[1]: *** [runtime/CMakeFiles/vimball-tags.dir/all] Error 2
make[1]: Leaving directory `/usr/src/RPM/BUILD/neovim-0.2.0/BUILD'
make: *** [all] Error 2
make: Leaving directory `/usr/src/RPM/BUILD/neovim-0.2.0/BUILD'
```

### PR DESCRIPTION
**Pull request by justinmk**
If procfs is missing libuv cannot find the exe path, so fallback to
arg[0].

Closes #6734

### CODE REVIEW

**Comment by jamessan** [comment]
What about porting [findYourself](https://github.com/vim/vim/blob/b5aedf3e228d35821591da9ae8501b61cf2e264c/src/workshop.c#L1760-L1821) as a fallback?

**Comment by justinmk** [comment]
Ported `findYourself`.  Thanks for the suggestion @jamessan

File: src/nvim/path.c [hunk 1]
```diff
@@ -2237,3 +2237,40 @@ int path_is_absolute_path(const char_u *fname)
...
+    xstrlcpy(buf, argv0, bufsize);
+  } else if (argv0[0] == '.' || strchr(argv0, PATHSEP)) {
+    if (os_dirname((char_u *)buf, MAXPATHL) != OK) {
+      buf[0] = NUL;
+    }
+    xstrlcat(buf, PATHSEPSTR, bufsize);
+    xstrlcat(buf, argv0, bufsize);
+  } else {
+    char  *pathbuf = xstrdup(path);
+    char  *runpath = xmalloc(MAXPATHL);
```
Comment (ZyX-I) [inline src/nvim/path.c]: Wondering why not use NameBuff or a `static char` array instead.

File: src/nvim/path.c [hunk 2]
```diff
@@ -2237,3 +2237,40 @@ int path_is_absolute_path(const char_u *fname)
...
+    path = strtok(pathbuf, ENV_SEPSTR);
```
Comment (ZyX-I) [inline src/nvim/path.c]: BTW, there is `vim_colon_env_iter` which does not require allocating new string. Just need a bit of refactoring to use a parameter in place of hardcoded `:`.

Comment (ZyX-I) [inline src/nvim/path.c]: Also unlike `strtok` `env_iter` is thread-safe.

File: src/nvim/path.c [hunk 3]
```diff
@@ -2237,3 +2237,40 @@ int path_is_absolute_path(const char_u *fname)
...
+    do {
+      strcpy(runpath, path);
```
Comment (ZyX-I) [inline src/nvim/path.c]: Looks dangerous, who said path can’t exceed MAXPATHL?

File: test/unit/path_spec.lua [hunk 4]
```diff
@@ -293,6 +294,37 @@ describe('path_shorten_fname_if_possible', function()
   end)
 end)
 
+describe('path.c path_guess_exepath', function()
```
Comment (ZyX-I) [inline test/unit/path_spec.lua]: Also need to check that if one `$PATH` item exceeds MAXPATHL tests do not crash.

File: src/nvim/path.c [hunk 5]
```diff
@@ -2237,3 +2237,40 @@ int path_is_absolute_path(const char_u *fname)
...
+      xstrlcat(runpath, PATHSEPSTR, MAXPATHL);
+      xstrlcat(runpath, argv0, MAXPATHL);
+      if (os_can_exe((char_u *)runpath, NULL, false)) {
+        break;
+      }
+    } while ((path = strtok(NULL, ENV_SEPSTR)) != NULL);
+    xstrlcpy(buf, runpath, bufsize);
```
Comment (ZyX-I) [inline src/nvim/path.c]: Looks like if nothing found, it uses whatever was the last possible `nvim` location, regardless of whether that was found or not. Is this intentional?

Comment (justinmk) [inline src/nvim/path.c]: Any result by then will be nonsense, but it should probably default to argv0. Will update.

File: src/nvim/path.c [hunk 6]
```diff
@@ -2237,3 +2237,44 @@ int path_is_absolute_path(const char_u *fname)
...
+    }
+    xstrlcat(buf, PATHSEPSTR, bufsize);
+    xstrlcat(buf, argv0, bufsize);
+  } else {
+    const void *iter = NULL;
+    do {
+      const char *dir;
+      size_t dir_len;
+      iter = vim_env_iter(ENV_SEPCHAR, path, iter, &dir, &dir_len);
+      if (dir == NULL || dir_len == 0 || dir_len + 1 > sizeof(NameBuff)) {
```
Comment (ZyX-I) [inline src/nvim/path.c]: I believe last part should not yield `break`, but `continue`.

File: test/unit/path_spec.lua [hunk 7]
```diff
@@ -293,6 +293,60 @@ describe('path_shorten_fname_if_possible', function()
...
+  local cwd = lfs.currentdir()
+
+  for _,name in ipairs({'./nvim', '.nvim', 'foo/nvim'}) do
+    itp('"'..name..'" returns name catenated with CWD', function()
+      local name = to_cstr(name)
```
Comment (ZyX-I) [inline test/unit/path_spec.lua]: BTW, most of time you may remove `to_cstr` and just use lua strings. Luajit will handle necessary conversions.

Comment (justinmk) [inline test/unit/path_spec.lua]: In this case luajit complains:

    test/unit/path_spec.lua:303: bad argument #1 to 'path_guess_exepath' (cannot convert 'string' to 'char *')'

**Comment by justinmk** [comment]
RDY after CI.

File: test/unit/path_spec.lua (hunk 7 above)
Comment (ZyX-I) [inline test/unit/path_spec.lua]: @justinmk Strange, see e.g. https://github.com/neovim/neovim/blob/17531ed0825c1bf5e24e78ae94020fb08ca06013/test/unit/eval/typval_spec.lua#L1636. Maybe it needs `const` to do automatic conversion?

File: src/nvim/path.c [hunk 8]
```diff
@@ -2237,3 +2237,50 @@ int path_is_absolute_path(const char_u *fname)
...
 }
+
+/// Builds a full path from an invocation name `argv0`, based on heuristics.
+///
+/// @param[in]  argv0     Name by which Nvim was invoked.
+/// @param[out] buf       Guessed full path to `argv0`.
+/// @param[in]  bufsize   Size of `buf`.
+///
+/// @see os_exepath
+void path_guess_exepath(char *argv0, char *buf, size_t bufsize)
```
Comment (ZyX-I) [inline src/nvim/path.c]: `argv0` is definitely `const char *` here, it is not being mutated.

Comment (ZyX-I) [inline src/nvim/path.c]: Would suggest to try removing `to_cstr` after making this `const`.

Comment (justinmk) [inline src/nvim/path.c]: @ZyX-I That was it! Thanks.

File: test/unit/path_spec.lua [hunk 9]
```diff
@@ -293,6 +293,59 @@ describe('path_shorten_fname_if_possible', function()
...
+    local name = 'cat'  -- Some executable in $PATH.
+    local bufsize = 255
+    local buf = cstr(bufsize, '')
+    local insane_path = orig_path_env..':'..(("x/"):rep(4097))
+
+    cimp.os_setenv(to_cstr('PATH'), to_cstr(insane_path), true)
+    cimp.path_guess_exepath(name, buf, bufsize)
+    eq("bin/cat", string.match(ffi.string(buf), "....cat$"))
+
+    -- Restore $PATH.
```
Comment (ZyX-I) [inline test/unit/path_spec.lua]: `itp` tests are run in a separate process, no need to bother restoring anything.

File: test/unit/path_spec.lua [hunk 10]
```diff
@@ -293,6 +293,59 @@ describe('path_shorten_fname_if_possible', function()
...
+  end)
+
+  itp('does not crash if $PATH item exceeds MAXPATHL', function()
+    local orig_path_env = os.getenv('PATH')
+    local name = 'cat'  -- Some executable in $PATH.
+    local bufsize = 255
+    local buf = cstr(bufsize, '')
+    local insane_path = orig_path_env..':'..(("x/"):rep(4097))
+
+    cimp.os_setenv(to_cstr('PATH'), to_cstr(insane_path), true)
```
Comment (ZyX-I) [inline test/unit/path_spec.lua]: BTW, os_setenv uses `const char *` as well.

File: test/unit/path_spec.lua (hunk 9 above)
Comment (justinmk) [inline test/unit/path_spec.lua]: I know, but I did this to be explicit.

File: test/unit/path_spec.lua [hunk 11]
```diff
@@ -293,6 +293,59 @@ describe('path_shorten_fname_if_possible', function()
...
+    cimp.path_guess_exepath(name, buf, bufsize)
+    eq("bin/cat", string.match(ffi.string(buf), "....cat$"))
```
Comment (ZyX-I) [inline test/unit/path_spec.lua]: Should not it be

```lua
eq('bin/' .. name, ffi.string(buf):sub(-#('bin/' .. name), -1))
```

? I mean, without this `name` variable is pretty useless. Same in a test below.

**Review by ZyX-I — APPROVED** [review]

**Pull request merged**